lib/ts/authUtils.ts

```diff
@@ -0,0 +1,280 @@
+                });
+            }
+        } else {
+            // In this case the app will try to link the session user and the authenticating user after auth
+            // so we consider this to be an MFA factor setup (if the MFA recipe is initialized)
+            // and we check if factor setup is allowed accordingly
+            let caughtSetupFactorError;
+            if (mfaInstance !== undefined) {
+                const factorsAlreadySetUp = await MultiFactorAuth.getFactorsSetupForUser(
+                    session!.getUserId(userContext),
+                    userContext
+                );
+                // In all apis besides passwordless createCode, we know exactly which factor we are signing into, so in those cases,
+                // this is basically just checking if the single factor is allowed to be setup or not.
+                // For createCode (if the FE didn't pass the factor id exactly, which it should for MFA),
+                // we filter whatever is allowed. If any of them are allowed, createCode can happen.
+                // The filtered list can be used to select email templates. As an example:
+                // If the flowType for passwordless is USER_INPUT_CODE_AND_MAGIC_LINK and but only the otp-email factor is allowed to be set up
+                // then we do not want to include a link in the email.
+                for (const id of factorIds) {
+                    try {
+                        if (!factorsAlreadySetUp.includes(id)) {
+                            await MultiFactorAuth.assertAllowedToSetupFactorElseThrowInvalidClaimError(
+                                session!,
+                                id,
+                                userContext
+                            );
+                        }
+                        // we add it to the valid factor ids list since it is either already set up or allowed to be set up
+                        validFactorIds.push(id);
+                    } catch (err) {
+                        caughtSetupFactorError = err;
+                    }
+                }
+                if (validFactorIds.length === 0) {
+                    throw caughtSetupFactorError;
+                }
+            }
+        }
+        // If this is a sign up we  check that the sign up is allowed
+        // for sign ins, this is checked after the credentials have been verified
+        if (isSignUp) {
+            if (
+                !(await accountLinkingInstance.isSignUpAllowed({
+                    newUser: accountInfo,
+                    isVerified,
+                    tenantId,
+                    session,
+                    userContext,
+                }))
+            ) {
+                return { status: "SIGN_IN_UP_NOT_ALLOWED" };
+            }
+        }
+
+        // If nothing failed, we return OK
+        return { status: "OK", validFactorIds, isFirstFactor: !wantToLinkSessionUser };
+    },
+
+    postAuthChecks: async function ({
```

Comment:
https://supertokens.slack.com/archives/D02B887A2MQ/p1706808245781919?thread_ts=1706804574.214459&cid=D02B887A2MQ

Reply:
The above change is required to the logic overall